torch/distributed/_composable/replicate.py

```diff
@@ -23,62 +25,49 @@ def replicate(
         >>> replicate(module)
     """
     torch._C._log_api_usage_once("torch.distributed.replicate")
-    _ReplicateState().mark_module(module, **kwargs)
+    _ReplicateState(ignored_modules=ignored_modules).mark_module(module, **kwargs)
     return module
 
 
-def _can_compose(module: nn.Module) -> bool:
-    r"""Check if module is composable for `replicate` API."""
-    return "fully_shard" not in _get_registry(module)
-
-
 class _ReplicateState:
-    def __init__(self) -> None:
+    def __init__(self, ignored_modules) -> None:
         self.module = None
         self.has_initialized: bool = False
         self._param_list: nn.ParameterList = nn.ParameterList()
         self.kwargs: dict = {}
+        self.ignored_modules = (
+            set(ignored_modules) if ignored_modules is not None else set()
+        )
 
     def mark_module(self, module: nn.Module, **kwargs) -> None:
-        if not _can_compose(module):
-            raise AssertionError(
-                "Cannot apply `replicate()` on a Module already managed by `fully_shard`"
-            )
         self.module = module
         replicate.state(module)._params_collected = False
         module.register_forward_pre_hook(self.forward_pre_hook)
         # TODO(@yhcharles): fix type error
         module.register_forward_hook(self.forward_post_hook)  # type: ignore[arg-type]
         self.kwargs = kwargs
 
-    def _recursive_collect_params(self, module: nn.Module) -> None:
-        # skip if managed by other APIs
-        if not _can_compose(module):
-            return
+    def _collect_params(self, module: nn.Module) -> None:
 
-        # skip if module parameters already collected
-        replicate_state = replicate.state(module)
-        # if replicate_state is None, `module` is a child module that has not been explicitly
-        # tagged as replicate().
-        if replicate_state is not None:
-            if hasattr(replicate_state, "_params_collected"):
-                if replicate.state(module)._params_collected:
-                    return
-                replicate.state(module)._params_collected = True
-
-        self._param_list.extend(
-            param for param in module.parameters(recurse=False) if param.requires_grad
-        )
-        for child in module.children():
-            self._recursive_collect_params(child)
+        if module in self.ignored_modules:
+            return  # if module A is ignored, all of A's children are also ignored.
+
+        ignored_params: Set[torch.nn.Parameter] = {
+            p for m in self.ignored_modules for p in m.parameters()
```

Comment:
The original idea to make `replicate` to be composable is: assuming we use it on a given Module, then it will recursively check all the sub-modules, apply data parallel on the sub-modules only when the sub-module is not marked by another composable API (since all composable APIs follows `contract`, they are supposed to be able to know each other).
It seems this change removes this logic, and makes `replicate` not composable. Is there any reason or it's my misunderstanding?
cc @mrshenli

Reply:
IIUC this was not completely the previous behavior.

Previously, `_can_compose` only checked for fully_shard, so trec_shard would get through. Thus when I tried to do a top-level replicate, replicate would try to mark embedding modules as data parallel which is not desired. 

The right fix would be to add the registry check for trec_shard as exists for _can_compose. However, trec_shard does not seem to add to the registry when it is wrapping an embedding submodule (only adds it for the top level). Once we fix this in trec_shard we may be able to get rid of ignore_modules. 

We should also make sure fully_shard can get rid of it as well then.

Reply:
Yes, the previous implementation was not complete -- only `fully_shard` and `replicate` run the check and thus are composable with each other. And also agree that the right fix is to enable the check for trec_shard.

Q:
- If we plan to add this check to trec_shard, then why removing it from `replicate`? It seems adding `ignored_modules` has not conflicts with the check.
- Will removing this check break the existing composability of `fully_shard` and `replicate`?